### PR TITLE
Uplift ArticlePage to TypeScript + `Article` type updates

### PR DESCRIPTION
**Pull request by amoore108**
Overall changes
======
- Uplifts `ArticlePage` and `TopStories` components and related tests to TypeScript
- Updates type for `Article` to better reflect data model and update references across codebase where needed
- Updates various utilities used to extract metadata to TypeScript
- Removes `analyticsUtils/article` utilities as they are no longer used. The `getPromoHeadline` function is the same as the existing `getHeadline` function that the `ArticlePage` component uses
- Removes `PagePromoSections/RelatedContentSection` as its since been superseded by `components/RelatedContentSection`
- Removes Ramda usage in `ArticlePage` and `MediaArticlePage` components

Testing
======
1. Ensure tests pass and Article and MediaArticle pages continue rendering content as before
2. There should be no visual regression

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

### CODE REVIEW

File: src/app/components/Byline/utilities/bylineExtractor/index.tsx [hunk 1]
```diff
@@ -1,18 +1,17 @@
+import {
+  OptimoBylineContributorBlock,
+  OptimoBylineContributorMetadataBlock,
+} from '#app/models/types/optimo';
 import pathOr from 'ramda/src/pathOr';
```
Comment (karinathomasbbc) [inline src/app/components/Byline/utilities/bylineExtractor/index.tsx:5]: Did we manage to get rid of all references to `pathOr`?

File: src/app/pages/ArticlePage/index.test.tsx [hunk 2]
```diff
@@ -113,8 +114,7 @@ const Context = ({
 beforeEach(() => {
   process.env.SIMORGH_ICHEF_BASE_URL = 'https://ichef.test.bbci.co.uk';
 
-  fetch.resetMocks();
-  ATIAnalytics.mockImplementation(
+  (ATIAnalytics as jest.Mock).mockImplementation(
```
Comment (karinathomasbbc) [inline src/app/pages/ArticlePage/index.test.tsx:121]: Will these mocks need to be reset in a `beforeEach` or `afterEach`?

**Review by karinathomasbbc — APPROVED** [review]

File: src/app/components/Byline/utilities/bylineExtractor/index.tsx (hunk 1 above)
Comment (amoore108) [reply src/app/components/Byline/utilities/bylineExtractor/index.tsx:5]: Theres still a reference in this file: https://github.com/bbc/simorgh/blob/6bb9d21540f7a645e89bf242624ba0dbc7e44693/src/app/components/Byline/utilities/bylineExtractor/index.tsx#L25

Its a tricky one to move off `pathOr`, as it constructs an array path to a nested block.model, like:
`pathOr('', ['some', 'long', 'path', here'], block)`.

Not sure of the best way of doing that outside of using `.reduce` perhaps?

File: src/app/pages/ArticlePage/index.test.tsx (hunk 2 above)
Comment (amoore108) [reply src/app/pages/ArticlePage/index.test.tsx:121]: Tests further down mock the implementation again https://github.com/bbc/simorgh/blob/afdc1fb70016ab77ebe95be8f3a7381f435413e1/src/app/pages/ArticlePage/index.test.tsx#L798 so I think its probably ok to not reset?

File: src/app/components/Byline/utilities/bylineExtractor/index.tsx (hunk 1 above)
Comment (karinathomasbbc) [reply src/app/components/Byline/utilities/bylineExtractor/index.tsx:5]: Ah I see, let's leave it. Yes, it's a nightmare to move away from, as there's a risk of accidentally breaking stuff. I am pleased to see you have managed to get rid of it in other places as part of this work though! Thanks! 🚀

File: src/app/pages/ArticlePage/index.test.tsx (hunk 2 above)
Comment (karinathomasbbc) [reply src/app/pages/ArticlePage/index.test.tsx:121]: Hmm. What if we were to reset it in an `afterAll` which means that this test cleans up after itself?

**Review by emilysaffron — APPROVED** [review]

**Pull request merged**